### PR TITLE
narwhal: require authorization for primary-to-worker and worker-to-primary interfaces

### PR DESCRIPTION
**Pull request by bmwill**
Require authorization for primary-to-worker and worker-to-primary
    interfaces by wrapping them in a `RequireAuthorizationLayer` which
    enforces that requests to the primary-to-worker service on a worker come
    from that worker's primary and that requests to the worker-to-primary
    service on a primary come from that primary's workers.

### CODE REVIEW

**Comment by andll** [comment]
Note that we do have primary-to-worker RPCs in subscriber (namely, RequestBatch) that come to worker from other primaries

**Comment by bmwill** [comment]
@andll it seems there are two "RequestBatch(s)" Rpcs, one in primary-to-worker and one in worker-to-worker. Maybe we want to utilize the one in the worker-to-worker interface? Either that or we maybe will want to break things out into difference services (maybe with better names) so we can have more fine-grained control

**Comment by andll** [comment]
Is it possible to have this authorization on per-RPC level?

We have a place where primary need to go get batch from other authority. Currently it issues RequestBatch RPC to the "remote" worker (e.g. worker of other authority).

I guess there is an alternative of just send rpc to own worker that will in tern send RPC to other worker - but it seem to be a strange choice (why do two nested RPC calls where we can do 1 instead?)

**Comment by bmwill** [comment]
> Is it possible to have this authorization on per-RPC level?

So that's challenging today and I don't know if there would be a good way to do that within a single service in a generic way, its much easier to do it at the level of a whole service (or group of rpcs).

> I guess there is an alternative of just send rpc to own worker that will in tern send RPC to other worker - but it seem to be a strange choice (why do two nested RPC calls where we can do 1 instead?)

Oh we don't even have to do that, If this particular rpc is something that we want other primaries to be able to call on any worker then we can just move this particular rpc to another service which we don't enforce auth on.

**Review by asonnino — APPROVED** [review]
LGMT

File: narwhal/primary/src/state_handler.rs [hunk 1]
```diff
@@ -68,7 +76,53 @@ impl StateHandler {
...
+        let our_workers = self
+            .worker_cache
+            .load()
+            .our_workers(&self.name)
+            .unwrap()
+            .into_iter()
+            .map(|info| info.name)
+            .collect();
+
+        self.network.unreliable_broadcast(our_workers, &message);
```
[the remark this answers is not in the document]
Comment (bmwill) [reply narwhal/primary/src/state_handler.rs:122]: That's a good question, I didn't think very much about which one to use so I picked the "cheaper" of the two. One potential problem could be if we waited for success with a reliable one we may have some tests that hang that only spawn primaries and not workers (Maybe we configure those tests to have zero workers, does that even make sense in narwhal to not have a worker?)

**Pull request merged**